storages/backends/azure_storage.py

```diff
@@ -1,132 +1,170 @@
-        self._block_list.append(BlobBlock(block_id))
-        self._last_commit_pos = self.file.tell()
-
-    def _flush_all_buffers(self):
-        """
-        Flushes the write buffer.
-        """
-        pos_before_flush = self.file.tell()
-        while self._needs_flush(pos_before_flush):
-            self._flush_buffer()
-        self.file.seek(pos_before_flush)
+        return super(AzureStorageFile, self).write(force_bytes(content))
 
     def close(self):
+        if self._file is None:
+            return
         if self._is_dirty:
-            self._flush_buffer()
-            self._storage.connection.put_block_list(self._storage.azure_container, self._name, self._block_list)
-        if self._file is not None:
-            self._file.close()
-            self._file = None
+            self._file.seek(0)
+            self._storage.save(self.name, self._file)
+            self._is_dirty = False
+        self._file.close()
+        self._file = None
+
+
+def _content_type(name, content):
+    try:
+        return content.file.content_type
+    except AttributeError:
+        pass
+    try:
+        return content.content_type
+    except AttributeError:
+        pass
+    return None
+
+
+def _get_valid_path(s):
+    # A blob name:
+    #   * must not end with dot or slash
+    #   * can contain any character
+    #   * must escape URL reserved characters
+    # We allow a subset of this to avoid
+    # illegal file names. We must ensure it is idempotent.
+    s = force_text(s).strip().replace(' ', '_')
+    s = re.sub(r'(?u)[^-\w./]', '', s)
+    s = s.replace('//', '/').strip('./')
+    if len(s) > _AZURE_NAME_MAX_LEN:
+        raise ValueError(
+            "File name max len is %d" % _AZURE_NAME_MAX_LEN)
+    if not len(s):
+        raise ValueError(
+            "File name must contain one or more "
+            "printable characters")
+    if s.count('/') > 256:
+        raise ValueError(
+            "File name must not contain "
+            "more than 256 slashes")
+    return s
+
+
+# Max len according to azure's docs
+_AZURE_NAME_MAX_LEN = 1024
 
 
 @deconstructible
 class AzureStorage(Storage):
+
     account_name = setting("AZURE_ACCOUNT_NAME")
     account_key = setting("AZURE_ACCOUNT_KEY")
     azure_container = setting("AZURE_CONTAINER")
     azure_ssl = setting("AZURE_SSL")
-    max_memory_size = setting('AZURE_BLOB_MAX_MEMORY_SIZE', 0)
-    buffer_size = setting('AZURE_FILE_BUFFER_SIZE', 4194304)
-
-    def __init__(self, *args, **kwargs):
-        super(AzureStorage, self).__init__(*args, **kwargs)
+    upload_max_conn = setting("AZURE_UPLOAD_MAX_CONN", 2)
+    timeout = setting('AZURE_CONNECTION_TIMEOUT_SECS', 20)
+    max_memory_size = setting('AZURE_BLOB_MAX_MEMORY_SIZE', 2*1024*1024)
+    expiration_secs = setting('AZURE_URL_EXPIRATION_SECS')
+    overwrite_files = setting('AZURE_OVERWRITE_FILES', True)
+    _location = setting('AZURE_LOCATION', '')
```

Comment:
Why `_location` and not `location`?

Reply:
I guess I thought it'd be more clear/clean having the class var as "private" and the attribute public. But now that I know users will actually subclass this, it'd be better to leave it public. Good catch! thanks! I'll update the `nitely_azure` branch.